src/NHibernate/Type/NullableType.cs

```diff
@@ -152,6 +153,15 @@ public sealed override void NullSafeSet(DbCommand st, object value, int index, I
 				// I definitely need to look into this more...
 				st.Parameters[index].Value = DBNull.Value;
 			}
+			else if (value is DelayedPostInsertIdentifier)
+			{
+				if (IsDebugEnabled)
+				{
+					Log.Debug("binding 0 to parameter: {0}", index);
+				}
+
+				st.Parameters[index].Value = 0;
```

Comment:
This assumes the identifier is compatible with an `int`, and this assumes `0` is not a valid value for it. This will be wrong in many cases.

Reply:
DelayedPostInsertIdentifier exists only for auto-increment fields, so I believe that the value 0 in this case is correct

Reply:
I compared this behavior with entity framework core 3 comparison by reference for an entity that is not yet stored in the table returns exactly parameter 0 in all cases

Reply:
What if `entity.Id` is not 0 for transient entity and for instance it's -1?
Your fix is incorrect for the following SQL Server table:
```SQL
CREATE TABLE Entity
(  
Id int IDENTITY(0,1), -- start from 0
Name varchar (20) 
 );  
```